timemachine/fe/dummy.py

```diff
@@ -581,3 +612,171 @@ def generate_optimal_dg_ag_pairs(mol, core, bond_idxs):
+    are due to either missing interactions, or incompatible parameters. This does not prune interactions in the
+    dummy-dummy regions since they are required for correctness. While core-core interactions can be pruned, we
+    choose to leave them untouched for now.
+
+    0) All bond terms are untouched.
+    1) All angles i-j-k must have bond terms i-j and j-k present, and with k > 10 and b > 0.2 angstrom on said bonds
+    2) All torsions i-j-k-l must have angle terms i-j-k and j-k-l present, and with k > 10 and
+        abs(angle - pi) > 0.05 rad and abs(angle - 0) > 0.05. In addition, bonded terms i-j, j-k,
+        and k-l must be present, and with k > 10 and b > 0.2 angstrom
+
+    Note that this does not *strictly* guarantee that the system is stable. For example, ring systems do not have
+    angles/bonds set to their equilibrium values:
+
+                D
+               /
+        0--1--2
+        |  |  |
+        3--4--5
+
+    The torsion term 0-1-2-D is numerically unstable, even though the angles and bonds are well-defined. Note that the
+    0-1-2 angle term is not defined in the forcefield as having an equilibrium angle of zero. One additional sanity check
+    that we can probably add later on is to also check the *initial* geometry's bond/angle length parameters.
+
+    Conversely, an interaction may still be stable even if certain sub-component interactions are missing. For example,
+    improper torsions are missing bonds and angles in the trefoil definitions, but the geometry is well defined nevertheless.
+
+    Parameters
+    ----------
+    core: list of int
+        core indices
+
+    bond_idxs: list of 2-tuples
+        (i,j) tuples of the bonded terms
+
+    bond_params: list of 2-tuples
+        (k, b) force constant, bond length
+
+    angle_idxs: list of 3-tuples
+        (i,j,k) tuples where the angle is defined between i-j-k
+
+    angle_params: list of 2-tuples
+        (k, a) force constant, angle in radians
+
+    torsion_idxs: list of 4-tuples
+        (i,j,k,l) tuples where the torsion is defined over i-j-k-l
+
+    torsion_idxs: list of 3-tuples
+        (k, p, n) force constant, period, phase
+
+    Returns
+    -------
+    keep_angle_flags, keep_torsion_flags
+        boolean flags indicating which angles and torsions we keep
+
+    """
+
+    assert len(bond_idxs) == len(bond_params)
+    assert len(angle_idxs) == len(angle_params)
+    assert len(torsion_idxs) == len(torsion_params)
+
+    bond_kv = {}
+    for ij, params in zip(bond_idxs, bond_params):
+        bond_kv[ordered_tuple(ij)] = params
+
+    keep_angle_flags = []
+
+    def ixn_is_all_core_or_all_dummy(atom_idxs):
+        all_core = np.all([a in core for a in atom_idxs])
+        all_dummy = np.all([a not in core for a in atom_idxs])
+
+        return all_core or all_dummy
+
+    for (i, j, k), params in zip(angle_idxs, angle_params):
+        b0 = ordered_tuple((i, j))
```

Comment:
The name `ordered_tuple` is a bit confusing to me, given what this function does (i.e. it sounds like it might sort the tuple, but it actually only looks at the first 2 elements and reverses the order if the first is larger than the second?). Would something like `canonicalize_ixn_tuple` or similar make sense?

Reply:
+1 for renaming to `canonicalize_` or similar

------

Tangent: Can we confirm that the implementation of `canonicalize_` should be the same for proper torsions and improper torsions?

In the force field handlers, `ReversibleBondHandler` applies a similar 2-fold symmetry to all valence terms except for impropers, which instead use this 3-fold symmetry:

https://github.com/proteneer/timemachine/blob/7c682225fd72deba70ad8003820061ca19dce63c/timemachine/ff/handlers/bonded.py#L211-L214

(If `canonicalize` should mirror this distinction, `canonicalize_` could be implemented as `min(permutations)` where `reversible_permutations(atom_idxs) = [atom_idxs, atom_idxs[::-1]]  # 2 permutations`, `improper_permutations(atom_idxs) = ...  # 3 permutations`. If `canonicalize` does not need to mirror this distinction, an inline comment is warranted.)

Reply:
1.  Will rename to canonicalize - agreed that its better
2. Both propers and impropers get converted in to periodic torsions - which are then canonicalized. The requirement for canonicalization is that U(i,j,k,l)  = U(l,k,j,i). The 3-fold symmetry here is only for the generation of the 4-component terms. If anything, the improper hander should be sorting the generated trefoil terms with `sort_tuple`.

Reply:
#658 address this more broadly.